### PR TITLE
[npud] Introduce test framework

### PR DESCRIPTION
**Pull request by jyoungyun**
This commit introduces test framework on npud.
It tests Server and Signal classes.

Signed-off-by: Jiyoung Yun <jy910.yun@samsung.com>

draft: #9767

### CODE REVIEW

File: runtime/service/npud/tests/core/Server.test.cc
Comment (glistening) [inline runtime/service/npud/tests/core/Server.test.cc:41]: @jyoungyun Why do you create thread for every test case?

**Comment by chunseoklee** [comment]
Q.  No problem in server(due to singleton) when two test suites run in parallel ? (I am not sure that two test suites run at the same time.)

File: runtime/service/npud/tests/core/Server.test.cc
Comment (jyoungyun) [reply runtime/service/npud/tests/core/Server.test.cc:41]: > First, I don't know how `Server` will be used in usual scenario and what is your expectation for each API.

The intention of Server test case is to check whether the server is running or not.

> I think it can be done without thread.

If not implementing using threads, the test main function is stuck because the Server holds the process until an event occurs.

**Comment by jyoungyun** [comment]
> Q.  No problem in server(due to singleton) when two test suites run in parallel ? (I am not sure that two test suites run at the same time.)  

As you told, this test case should be executed in serialization. Because the server was implemented in Singleton, this is only one instance, which is why the test case did not work well. Does ONE Runtime's gtest run in parallel ?

**Comment by chunseoklee** [comment]
>Does ONE Runtime's gtest run in parallel ?

I cannot find any documentation about gtest's parallelism. But, https://github.com/google/gtest-parallel implies that vanilla gtest runs in sequencial ?

File: runtime/service/npud/tests/core/Server.test.cc
Comment (glistening) [reply runtime/service/npud/tests/core/Server.test.cc:41]: @jyoungyun Frankly speaking, I still don't understand why it needs thread. Is it necessary just for making a test? Maybe I need to read a whole code. Please proceed without my approval.

Comment (jyoungyun) [reply runtime/service/npud/tests/core/Server.test.cc:41]: Yes, this thread just needs for test environment. To test server functions, I need to ask some commands to server from client. In this test case, the child thread is a server and the main thread is a client. I'm not sure this implement is right approach for server tests. However, when searching for various data, I wasn't find a general way to test Server, so it is implemented so that it can be tested using gtest. If there is a better way, I will revise it later. :)

**Comment by jyoungyun** [comment]
> I cannot find any documentation about gtest's parallelism. But, https://github.com/google/gtest-parallel implies that vanilla gtest runs in sequencial ? 

I think so. When we need to consider the parallel test, I will update these test cases. :)

**Review by hseok-oh — APPROVED** [review]

File: runtime/service/npud/tests/core/Server.test.cc
Comment (glistening) [reply runtime/service/npud/tests/core/Server.test.cc:41]: > I wasn't find a general way to test Server, 

As I understand, npud will use d-bus and glib mainloop.

Then, I don't think testcase cover the actual case.

First, server and client will be in different processes. Client is not likely to use `Server::isRunning` (How client can get the instance out of different process?) I would add timeout source and add the codes (isRunning, raise signal, ...).

Second, I guess you will revisit testcase after d-bus related code is added based on how server and client will actually work.

**Review by chunseoklee — APPROVED** [review]

**Comment by glistening** [comment]
If anyone is waiting for my approval, please find https://github.com/Samsung/ONE/pull/9828#discussion_r987569371.

**Pull request merged**
